pkg/csr/csr.go

```diff
@@ -0,0 +1,236 @@
+	if req == nil || csr == nil {
+		return false, errors.New("CSR or request should not be nil")
+	}
+
+	// Check groups, we need at least: system:nodes, system:authenticated
+	if len(req.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR contains too few groups: %s", req.Name)
+	}
+	groupSet := sets.NewString(req.Spec.Groups...)
+	if !groupSet.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR does not contain required groups: %s", req.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !hasExactUsages(req, kubeletServerUsages) {
+		return false, errors.Errorf("CSR does not contain required usages: %s", req.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR does not contain required subject organization: %s", req.Name)
+	}
+	return true, nil
+}
+
+// Approve approves a CSR by updating it's status conditions to true if it is a valid CSR
+func (c *CSRApprover) Approve(csr *certificatesv1.CertificateSigningRequest) error {
+
+	if validated, err := c.validateCSRContents(csr); !validated {
+		c.log.Info("CSR contents are invalid for approval by WMCO CSR Approver", "CSR", csr.Name)
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", csr.Name)
+	}
+
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:           certificatesv1.CertificateApproved,
+		Status:         "True",
+		Message:        "This CSR was approved by the WMCO CSR Approver",
+		LastUpdateTime: metav1.Now(),
+		Reason:         "WMCOApproved",
+	})
+
+	if _, err := c.k8sClientSet.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		csr.Name, csr, metav1.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "could not update conditions for CSR approval %s", csr.Name)
+	}
+	c.log.Info("CSR approved by WMCO CSR Approver", "CSR", csr.Name)
+	return nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func isNodeClientCert(csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, NodeUserPrefix) {
+		return false
+	}
+	return true
+}
+
+// hasExactUsages verifies if the required usages exist in the CSR spec
```

Comment:
I suggest naming it `hasUsages` given the input is a list of usages that you are checking against.